src/ui.rs

```diff
@@ -181,5 +178,31 @@ impl ViewState {
         let text = Text::from(text);
         let paragraph = Paragraph::new(text).block(right_block_down);
         f.render_widget(paragraph, right_chunks[1]);
+
+        if show_help {
+            let block = Block::bordered().title("Help");
+            let help_message = Paragraph::new("Key shortcuts:\n'h' for help\n's' to step one instruction\n'q' to quit").block(block);
+            let area = centered_rect(60, 20, size);
+            f.render_widget(Clear, area);
+            f.render_widget(help_message, area);
+        }
     }
 }
+
+
+/// From ratatui/examples/popup.rs
+fn centered_rect(percent_x: u16, percent_y: u16, r: Rect) -> Rect {
+    let popup_layout = Layout::vertical([
+        Constraint::Percentage((100 - percent_y) / 2),
+        Constraint::Percentage(percent_y),
+        Constraint::Percentage((100 - percent_y) / 2),
+    ])
+    .split(r);
+
+    Layout::horizontal([
+        Constraint::Percentage((100 - percent_x) / 2),
+        Constraint::Percentage(percent_x),
+        Constraint::Percentage((100 - percent_x) / 2),
+    ])
+    .split(popup_layout[1])[1]
```

Comment:
Web browser coding… probably a bit worse performance wise due to the additional vector creation here, but the usage feels nicer to me.
https://docs.rs/ratatui/0.26.2/ratatui/layout/struct.Layout.html#method.areas

```suggestion
    let [_, center, _] = Layout::vertical([
        Constraint::Percentage((100 - percent_y) / 2),
        Constraint::Percentage(percent_y),
        Constraint::Percentage((100 - percent_y) / 2),
    ])
    .areas(r);

    let [_, center, _] = Layout::horizontal([
        Constraint::Percentage((100 - percent_x) / 2),
        Constraint::Percentage(percent_x),
        Constraint::Percentage((100 - percent_x) / 2),
    ])
    .areas(center);

    center
```

Maybe something like this can be done even better…
Personally I wrote my own math as I find the Layout stuff somewhat confusing too…
https://github.com/EdJoPaTo/mqttui/blob/27dca3f099d6cb057870eefd31217703f232e207/src/interactive/clean_retained.rs#L31

Reply:
Thanks! Much more readable.